src/TArea.cpp

```diff
@@ -584,3 +606,372 @@ int TArea::createLabelId() const
+}
+
+// Takes a userData object and parses all its elements
+QMap<QString, QString> TArea::readJsonUserData(const QJsonObject& obj) const
+{
+    QMap<QString, QString> results;
+    if (obj.isEmpty()) {
+        // Skip doing anything more if there is nothing to do:
+        return results;
+    }
+
+    QStringList keys = obj.keys();
+    for (int i = 0, total = keys.count(); i < total; ++i) {
+        if (obj.value(keys.at(i)).isString()) {
+            results.insert(keys.at(i), obj.value(keys.at(i)).toString());
+        }
+    }
+    return results;
+}
+
+void TArea::writeJsonLabels(QJsonObject& obj) const
+{
+    if (mMapLabels.isEmpty()) {
+        // No labels in this area - so nothing to do
+        return;
+    }
+
+    QJsonArray labelArray;
+    QMapIterator<int, TMapLabel> itMapLabel(mMapLabels);
+    while (itMapLabel.hasNext()) {
+        itMapLabel.next();
+        writeJsonLabel(labelArray, itMapLabel.key(), &itMapLabel.value());
+        if (mpMap->incrementJsonProgressDialog(true, false, 1)) {
+            // Cancel has been hit - so give up straight away:
+            return;
+        }
+    }
+    QJsonValue labelsValue{labelArray};
+    obj.insert(scLABELS, labelsValue);
+}
+
+// obj is the (area) container that contains the label array
+void TArea::readJsonLabels(const QJsonObject& obj)
+{
+    const QJsonArray labelsArray = obj.value(scLABELS).toArray();
+
+    if (labelsArray.isEmpty()) {
+        // No labels at all in this area
+        return;
+    }
+
+    for (int index = 0, total = labelsArray.count(); index < total; ++index) {
+        readJsonLabel(labelsArray.at(index).toObject());
+        if (mpMap->incrementJsonProgressDialog(false, false, 1)) {
+            // Cancel has been hit - so give up straight away:
+            return;
+        }
+    }
+}
+
+void TArea::writeJsonLabel(QJsonArray& array, const int id, const TMapLabel* pLabel) const
+{
+    QJsonObject labelObj;
+
+    labelObj.insert(scID, static_cast<double>(id));
+
+    writeJson3DCoordinates(labelObj, scCOORDINATES, pLabel->pos);
+
+    writeJsonSize(labelObj, scSIZE, pLabel->size);
+
+    if (!(pLabel->text.isEmpty() || !pLabel->text.compare(tr("no text", "Default text if a label is created in mapper with no text")))) {
+        // Don't include the text if it is am image:
+        QJsonValue textValue{pLabel->text};
+        labelObj.insert(scTEXT, textValue);
+    }
+
+    if (!(pLabel->fgColor.red() == defaultLabelForeground.red()
+          && pLabel->fgColor.green() == defaultLabelForeground.green()
+          && pLabel->fgColor.blue() == defaultLabelForeground.blue()
+          && pLabel->bgColor.red() == defaultLabelBackground.red()
+          && pLabel->bgColor.red() == defaultLabelBackground.green()
+          && pLabel->bgColor.red() == defaultLabelBackground.blue())) {
+
+        // For an image the colors are not used and tend to be set to black, if
+        // so skip them. Ufortunately because of the way QColour s are assembled
```

Comment:
```suggestion
        // so skip them. Unfortunately because of the way QColours are assembled
```

Same with other cases

Reply:
:blush:

Reply:
Fixed in next commit.